tensorflow/contrib/layers/kernels/layer_norm_fused_op_gpu.cu.cc

```diff
@@ -0,0 +1,1455 @@
+                                        const T* __restrict__ gamma,
+                                        const T* __restrict__ beta,
+                                        T* __restrict__ output,
+                                        const int num_blocks) {
+  const int in_depth = args.depth;
+  const int slice_size = args.slice_size;
+  const int n_inputs = args.n_inputs;
+  const T epsilon = args.epsilon;
+
+  const int thread_warp_id = threadIdx.x % warpSize;
+
+  const int thread_slice_id = threadIdx.x % slice_size;
+
+  extern __shared__ __align__(sizeof(T)) unsigned char my_smem[];
+  T* mean_cache = (T*)my_smem;
+  T* std_cache = (T*)&my_smem[sizeof(T)];
+
+  const T i_n = static_cast<T>(1.0f) / static_cast<T>(in_depth);
+  T inp[mult];
+  T _gamma[mult];
+  T _beta[mult];
+  int thread_id[mult];
+
+  T sum;
+  T sqSum;
+  T mu;
+  T rstd;
+
+  UNROLL for (int m = 0; m < mult; m++) {
+    _gamma[m] = get_value<T>(gamma + thread_slice_id + m * blockDim.x,
+                             thread_slice_id + m * blockDim.x, in_depth);
+    _beta[m] = get_value<T>(beta + thread_slice_id + m * blockDim.x,
+                            thread_slice_id + m * blockDim.x, in_depth);
+  }
+
+  for (int bId = blockIdx.x; bId < num_blocks; bId += gridDim.x) {
+    sum = static_cast<T>(0.0f);
+    sqSum = static_cast<T>(0.0f);
+
+    if (thread_slice_id == 0) {
+      mean_cache[0] = static_cast<T>(0.0f);
+      std_cache[0] = static_cast<T>(0.0f);
+    }
+    __syncthreads();
+
+    UNROLL for (int m = 0; m < mult; m++) {
+      thread_id[m] = bId * in_depth + m * blockDim.x + thread_slice_id;
+    }
+
+    UNROLL for (int m = 0; m < mult; m++) {
+      // inp[m] = __ldg(input+thread_id[m]);
+      inp[m] = get_value<T>(input + thread_id[m],
+                            thread_slice_id + m * blockDim.x, in_depth);
+      // if(blockIdx.x==0&&threadIdx.x==0)
+      // printf("m:%d,bId:%d,tId:%d,inp:%f\n",m,bId,thread_id[m],inp[m]);
+    }
+
+    UNROLL for (int m = 0; m < mult; m++) { sum += inp[m] * i_n; }
+    for (int mask = warpSize / 2; mask > 0; mask /= 2) {
+      sum += __shfl_xor(sum, mask);
+    }
+    // if(blockIdx.x==0&&thread_warp_id==0)printf("wi:%d,sum:%f,psmu:%f\n",threadIdx.x/warpSize,
+    // sum,mean_cache[slice_id]);
+    if (thread_warp_id == 0) {
+      atomicAdd(&mean_cache[0], sum);
+    }
+    // if(blockIdx.x==0&&thread_warp_id==0)printf("--wi:%d,sum:%f,psmu:%f\n",threadIdx.x/warpSize,
+    // sum,mean_cache[slice_id]);
+    __syncthreads();
+
+    mu = mean_cache[0];
+    // if(threadIdx.x==0&&blockIdx.x==0)printf("mu:%f\n", mu);
+    UNROLL for (int m = 0; m < mult; m++) {
+      if (thread_slice_id + m * blockDim.x < in_depth)
+        sqSum += (inp[m] - mu) * (inp[m] - mu);
+    }
+    for (int mask = warpSize / 2; mask > 0; mask /= 2) {
+      sqSum += __shfl_xor(sqSum, mask);
+    }
+    if (thread_warp_id == 0) {
+      atomicAdd(&std_cache[0], sqSum);
```

Comment:
I think I used this approach originally, but it was actually slower (or similar, I am a bit fuzzy on this one since this is a month or two ago).I am not sure why, but my current hypothesis is that accumulating in the end required an additional `__syncthreads()`, and this `atomicAdd` is accessing shared memory with only one thread per warp accessing the location.